File: backend/app/api/v1/routes/transactions.py
import uuid
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query, Request, status
from sqlalchemy.orm import Session

from app.core.db import get_db
from app.dependencies.access import assert_loan_access
from app.dependencies.auth import get_current_user, require_admin
from app.models.loan import Loan
from app.models.transaction import Transaction, TransactionStatus
from app.models.user import User, UserRole
from app.schemas.transaction import (
    LoanTransactionSummary,
    PendingConfirmationItem,
    PendingConfirmationListResponse,
    TransactionCreate,
    TransactionListResponse,
    TransactionResponse,
    TransactionUpdate,
)
from app.services.loan import get_loan
from app.services.transaction import (
    confirm_transaction,
    create_transaction,
    fail_transaction,
    get_loan_transaction_summary,
    get_transaction,
    list_pending_confirmations,
    list_transactions,
    soft_delete_transaction,
    update_transaction,
)

router = APIRouter(prefix="/transactions", tags=["Transactions"])

# Audit lives in the service layer — see app/services/transaction.py.


def _assert_loan_in_user_scope(
    db: Session, loan_id: uuid.UUID, current_user: User
) -> Loan:
    loan = (
        db.query(Loan)
        .filter(Loan.id == loan_id, Loan.is_deleted.is_(False))
        .first()
    )
    if loan is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Loan not found"
        )
    assert_loan_access(loan, current_user, db)
    return loan


def _assert_transaction_in_user_scope(
    db: Session, transaction: Transaction, current_user: User
) -> None:
    _assert_loan_in_user_scope(db, transaction.loan_id, current_user)


# --------------------------------------------------
# CREATE
# --------------------------------------------------
@router.post(
    "/",
    response_model=TransactionResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Record a new payment",
)
def create_transaction_route(
    request: Request,
    payload: TransactionCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    _assert_loan_in_user_scope(db, payload.loan_id, current_user)
    # The frontend sends the dedupe key as an `Idempotency-Key` HTTP header
    # (client.ts interceptor), while the service dedupes on
    # payload.idempotency_key. Bridge the two so a retried POST carrying the
    # same key returns the existing row instead of creating a duplicate. An
    # explicit body value wins if a caller set both.
    if not payload.idempotency_key:
        header_key = request.headers.get("Idempotency-Key")
        if header_key:
            payload.idempotency_key = header_key[:64]
    try:
        return create_transaction(
            db=db, data=payload, created_by=current_user.id, request=request,
        )
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))


# --------------------------------------------------
# LIST
# --------------------------------------------------
@router.get(
    "/",
    response_model=TransactionListResponse,
    summary="List transactions with filters",
)
def list_all(
    loan_id: Optional[uuid.UUID] = Query(None),
    collected_by_id: Optional[uuid.UUID] = Query(None),
    status_filter: Optional[TransactionStatus] = Query(None, alias="status"),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    results, total, total_collected = list_transactions(
        db=db,
        loan_id=loan_id,
        collected_by_id=collected_by_id,
        status=status_filter,
        page=page,
        page_size=page_size,
        current_user_id=current_user.id,
        current_user_role=current_user.role,
    )
    return TransactionListResponse(
        total=total,
        page=page,
        page_size=page_size,
        total_collected=total_collected,
        results=results,
    )


# --------------------------------------------------
# PENDING CONFIRMATIONS WORKLIST (Collections & Actions)
# Registered before /{transaction_id} so the static path isn't shadowed.
# --------------------------------------------------
@router.get(
    "/pending-confirmations",
    response_model=PendingConfirmationListResponse,
    summary="Cross-loan list of payments awaiting confirmation",
)
def pending_confirmations(
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    sort_by: Optional[str] = Query(
        None,
        description="Sort column: amount | effective_payment_date | created_at | customer_name | loan",
    ),
    sort_order: Optional[str] = Query(None, description="asc | desc"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    scope = current_user.id if current_user.role == UserRole.EMPLOYEE else None
    rows, total, total_amount = list_pending_confirmations(
        db,
        assigned_employee_id=scope,
        page=page,
        page_size=page_size,
        sort_by=sort_by,
        sort_order=sort_order,
    )
    results = [
        PendingConfirmationItem(
            id=txn.id,
            loan_id=loan.id,
            loan_number=loan.loan_number,
            hp_number=loan.hp_number,
            customer_id=customer.id,
            customer_name=customer.full_name,
            customer_mobile=customer.mobile_number,
            amount=txn.amount,
            payment_mode=txn.payment_mode,
            effective_payment_date=txn.effective_payment_date,
            collected_by_id=txn.collected_by_id,
            created_at=txn.created_at,
            due_cycle_id=txn.due_cycle_id,
            cycle_number=cycle.cycle_number if cycle is not None else None,
            cycle_due_date=cycle.due_date if cycle is not None else None,
        )
        for txn, loan, customer, cycle in rows
    ]
    return PendingConfirmationListResponse(
        total=total,
        page=page,
        page_size=page_size,
        total_pending_amount=total_amount,
        results=results,
    )


# --------------------------------------------------
# GET BY ID
# --------------------------------------------------
@router.get(
    "/{transaction_id}",
    response_model=TransactionResponse,
    summary="Get a single transaction",
)
def get_one(
    transaction_id: uuid.UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    transaction = get_transaction(db, transaction_id)
    if not transaction:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Transaction not found"
        )
    _assert_transaction_in_user_scope(db, transaction, current_user)
    return transaction


# --------------------------------------------------
# LOAN SUMMARY
# --------------------------------------------------
@router.get(
    "/loan/{loan_id}/summary",
    response_model=LoanTransactionSummary,
    summary="Get outstanding balance for a loan",
)
def loan_summary(
    loan_id: uuid.UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    loan = _assert_loan_in_user_scope(db, loan_id, current_user)
    return get_loan_transaction_summary(db, loan)


# --------------------------------------------------
# CONFIRM TRANSACTION (Admin only)
# --------------------------------------------------
@router.post(
    "/{transaction_id}/confirm",
    response_model=TransactionResponse,
    summary="Confirm a pending transaction",
)
def confirm_transaction_route(
    request: Request,
    transaction_id: uuid.UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_admin),
):
    transaction = get_transaction(db, transaction_id)
    if not transaction:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Transaction not found"
        )
    try:
        return confirm_transaction(
            db=db,
            transaction=transaction,
            updated_by=current_user.id,
            request=request,
        )
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))


# --------------------------------------------------
# FAIL TRANSACTION (Admin only)
# --------------------------------------------------
@router.post(
    "/{transaction_id}/fail",
    response_model=TransactionResponse,
    summary="Mark a pending transaction as failed",
)
def fail_transaction_route(
    request: Request,
    transaction_id: uuid.UUID,
    reason: Optional[str] = Query(None, max_length=500),
    db: Session = Depends(get_db),
    current_user: User = Depends(require_admin),
):
    transaction = get_transaction(db, transaction_id)
    if not transaction:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Transaction not found"
        )
    try:
        return fail_transaction(
            db=db,
            transaction=transaction,
            updated_by=current_user.id,
            reason=reason,
            request=request,
        )
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))


# --------------------------------------------------
# UPDATE (edit-after-record)
# --------------------------------------------------
# Editing a transaction supports two cases:
#   1. PENDING/FAILED — anyone in scope of the loan (admin or assigned
#      employee) can correct a misclick before the txn is confirmed. This
#      mirrors the "the collector who recorded it can fix their own typo"
#      expectation; the loan-scope check below still applies.
#   2. SUCCESS — admin only. Editing a confirmed transaction recomputes
#      the cycle ledger (old + new cycle's total_received), so it's a
#      sensitive operation we lock behind ADMIN/SUPER_ADMIN.
@router.patch(
    "/{transaction_id}",
    response_model=TransactionResponse,
    summary="Edit a transaction (cycle, amount, mode, date, notes)",
)
def update_transaction_route(
    request: Request,
    transaction_id: uuid.UUID,
    payload: TransactionUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    transaction = get_transaction(db, transaction_id)
    if not transaction:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Transaction not found"
        )
    _assert_transaction_in_user_scope(db, transaction, current_user)
    if transaction.status == TransactionStatus.SUCCESS and current_user.role not in (
        UserRole.ADMIN,
        UserRole.SUPER_ADMIN,
    ):
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Only an admin can edit a confirmed transaction.",
        )
    try:
        return update_transaction(
            db=db,
            transaction=transaction,
            data=payload,
            updated_by=current_user.id,
            request=request,
        )
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))


# --------------------------------------------------
# SOFT DELETE (Admin only)
# --------------------------------------------------
@router.delete(
    "/{transaction_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Soft delete a transaction (FAILED only)",
)
def delete_transaction_route(
    request: Request,
    transaction_id: uuid.UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_admin),
):
    transaction = get_transaction(db, transaction_id)
    if not transaction:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Transaction not found"
        )
    try:
        soft_delete_transaction(
            db=db,
            transaction=transaction,
            deleted_by=current_user.id,
            request=request,
        )
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))

File: frontend/src/api/queries/transactions.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'

import { apiClient } from '@/api/client'
import { loanKeys } from '@/api/queries/loans'
import { dueCycleKeys } from '@/api/queries/dueCycles'
import type {
  PaymentMethod,
  PunctualityStatus,
  TransactionStatus,
  TransactionType,
} from '@/schemas/enums'

// --------------------------------------------------
// Lean read surface for the Loans module: the outstanding-balance summary
// the closure form needs. Full transactions CRUD belongs to a dedicated
// Transactions module. Monetary fields arrive as strings.
// --------------------------------------------------

export interface LoanTransactionSummary {
  loan_id: string
  principal: string
  total_payable: string
  total_paid: string
  total_pending: string
  outstanding: string
  transaction_count: number
}

export interface TransactionResponse {
  id: string
  loan_id: string
  amount: string
  payment_mode: PaymentMethod
  notes: string | null
  status: TransactionStatus
  transaction_type: TransactionType
  collected_by_id: string | null
  is_deleted: boolean
  created_by_id: string | null
  updated_by_id: string | null
  idempotency_key: string | null
  effective_payment_date: string
  punctuality_status: PunctualityStatus
  due_cycle_id: string | null
  created_at: string
  updated_at: string
}

export interface TransactionListResponse {
  total: number
  page: number
  page_size: number
  total_collected: string
  results: TransactionResponse[]
}

export const transactionKeys = {
  all: ['transactions'] as const,
  summary: (loanId: string) => [...transactionKeys.all, 'summary', loanId] as const,
  byLoan: (loanId: string) => [...transactionKeys.all, 'byLoan', loanId] as const,
  pendingConfirmations: () => [...transactionKeys.all, 'pendingConfirmations'] as const,
}

export function useLoanTransactions(loanId: string | undefined, enabled = true) {
  return useQuery({
    queryKey: transactionKeys.byLoan(loanId ?? ''),
    queryFn: async () => {
      const { data } = await apiClient.get<TransactionListResponse>('/transactions/', {
        params: { loan_id: loanId, page_size: 100 },
      })
      return data
    },
    enabled: !!loanId && enabled,
  })
}

// --------------------------------------------------
// Pending confirmations worklist — cross-loan PENDING transactions awaiting an
// admin's confirm/fail, enriched with loan + customer (+ cycle). Mirrors
// backend GET /transactions/pending-confirmations.
// --------------------------------------------------
export interface PendingConfirmationItem {
  id: string
  loan_id: string
  loan_number: string
  hp_number: string | null
  customer_id: string
  customer_name: string
  customer_mobile: string
  amount: string
  payment_mode: PaymentMethod
  effective_payment_date: string
  collected_by_id: string | null
  created_at: string
  due_cycle_id: string | null
  cycle_number: number | null
  cycle_due_date: string | null
}

export interface PendingConfirmationListResponse {
  total: number
  page: number
  page_size: number
  total_pending_amount: string
  results: PendingConfirmationItem[]
}

export type PendingSortField =
  | 'amount'
  | 'effective_payment_date'
  | 'created_at'
  | 'customer_name'
  | 'loan'

export function usePendingConfirmations(
  page: number,
  pageSize = 20,
  sort?: { sort_by?: PendingSortField; sort_order?: 'asc' | 'desc' },
) {
  return useQuery({
    queryKey: [
      ...transactionKeys.pendingConfirmations(),
      page,
      pageSize,
      sort?.sort_by ?? null,
      sort?.sort_order ?? null,
    ] as const,
    queryFn: async () => {
      const { data } = await apiClient.get<PendingConfirmationListResponse>(
        '/transactions/pending-confirmations',
        { params: { page, page_size: pageSize, ...sort } },
      )
      return data
    },
    placeholderData: (prev) => prev,
  })
}

export function useLoanSummary(loanId: string | undefined, enabled = true) {
  return useQuery({
    queryKey: transactionKeys.summary(loanId ?? ''),
    queryFn: async () => {
      const { data } = await apiClient.get<LoanTransactionSummary>(
        `/transactions/loan/${loanId}/summary`,
      )
      return data
    },
    enabled: !!loanId && enabled,
  })
}

// --------------------------------------------------
// Mutations — mirror backend/app/schemas/transaction.py. amount is a string.
// A created transaction is PENDING; an admin then confirms (→ SUCCESS, which
// updates the cycle's received total and can move the loan to AWAITING_CLOSURE)
// or fails it. The collection POST carries an auto Idempotency-Key (client.ts).
// --------------------------------------------------

export interface TransactionCreate {
  loan_id: string
  amount: string
  payment_mode: PaymentMethod
  // Required by the backend as of the cycle-required change — every payment
  // must land on a specific cycle to keep total_received in sync.
  due_cycle_id: string
  effective_payment_date?: string | null
  notes?: string | null
}

// Confirming/failing changes cycle totals and possibly loan status, so refresh
// the loan detail, schedule, and summary alongside the transaction list.
function invalidateLoanLedger(
  qc: ReturnType<typeof useQueryClient>,
  loanId: string,
) {
  qc.invalidateQueries({ queryKey: transactionKeys.byLoan(loanId) })
  qc.invalidateQueries({ queryKey: transactionKeys.summary(loanId) })
  // The confirmations worklist (and its tab badge) lists PENDING rows.
  qc.invalidateQueries({ queryKey: transactionKeys.pendingConfirmations() })
  // Covers both the per-loan schedule (byLoan) and the cross-loan worklist.
  qc.invalidateQueries({ queryKey: dueCycleKeys.all })
  qc.invalidateQueries({ queryKey: loanKeys.detail(loanId) })
}

export function useCreateTransaction(loanId: string) {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (payload: TransactionCreate) => {
      const { data } = await apiClient.post<TransactionResponse>('/transactions/', payload)
      return data
    },
    onSuccess: () => invalidateLoanLedger(qc, loanId),
  })
}

export function useConfirmTransaction(loanId: string) {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (transactionId: string) => {
      const { data } = await apiClient.post<TransactionResponse>(
        `/transactions/${transactionId}/confirm`,
      )
      return data
    },
    onSuccess: () => invalidateLoanLedger(qc, loanId),
  })
}

export function useFailTransaction(loanId: string) {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (transactionId: string) => {
      const { data } = await apiClient.post<TransactionResponse>(
        `/transactions/${transactionId}/fail`,
      )
      return data
    },
    onSuccess: () => invalidateLoanLedger(qc, loanId),
  })
}

// Worklist variants of confirm/fail: the loan id arrives per row rather than
// from a loan-scoped page, so it's part of the mutation variables.
export function useConfirmPendingTransaction() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (vars: { transactionId: string; loanId: string }) => {
      const { data } = await apiClient.post<TransactionResponse>(
        `/transactions/${vars.transactionId}/confirm`,
      )
      return data
    },
    onSuccess: (_data, vars) => invalidateLoanLedger(qc, vars.loanId),
  })
}

export function useFailPendingTransaction() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (vars: { transactionId: string; loanId: string }) => {
      const { data } = await apiClient.post<TransactionResponse>(
        `/transactions/${vars.transactionId}/fail`,
      )
      return data
    },
    onSuccess: (_data, vars) => invalidateLoanLedger(qc, vars.loanId),
  })
}

// Edit a transaction. All fields optional — only set fields are applied.
// PENDING/FAILED transactions are editable by any user in scope of the loan;
// SUCCESS transactions are editable by admins only (server-side gate). When
// `due_cycle_id` or `amount` changes on a SUCCESS row, the server recomputes
// total_received on both the old and new cycle.
// `notes: null` clears the existing note.
export interface TransactionUpdate {
  notes?: string | null
  due_cycle_id?: string | null
  amount?: string
  effective_payment_date?: string
  payment_mode?: PaymentMethod
}

export function useUpdateTransaction(loanId: string) {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (vars: { transactionId: string; payload: TransactionUpdate }) => {
      const { data } = await apiClient.patch<TransactionResponse>(
        `/transactions/${vars.transactionId}`,
        vars.payload,
      )
      return data
    },
    onSuccess: () => invalidateLoanLedger(qc, loanId),
  })
}

// Soft-delete a transaction (admin). Backend allows this for FAILED
// transactions only (400 otherwise); a wrong PENDING/SUCCESS txn must be
// failed first. Returns 204 — no body.
export function useDeleteTransaction(loanId: string) {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (transactionId: string) => {
      await apiClient.delete(`/transactions/${transactionId}`)
    },
    onSuccess: () => invalidateLoanLedger(qc, loanId),
  })
}
